File: 0010_Regular_Expression_Matching/Regular_Expression_Matching.py
# Check the C++ file for explaining comments, i cant understand what is happening in this code. I took help from AI
# TODO: Solve this problem again. Without the help of AI. 

class Solution:
    def f(self, i, j, p, s, dp):
        if (i < 0 and j < 0): return True
        if (i < 0 and j >= 0): return False

        if (i >= 0 and j < 0):
            while (i >= 0):
                if (p[i] != '*'): return False
                i -= 2
            return True

        if (dp[i][j] != -1): return dp[i][j]

        if (p[i] == s[j] or p[i] == '.'):
            dp[i][j] = self.f(i-1, j-1, p, s, dp)
            return dp[i][j]

        if (p[i] == '*'):
            zero = False
            if (i - 2 >= -1):
                zero = self.f(i - 2, j, p, s, dp)

            oneOrMore = False
            if (i - 1 >= 0 and j >= 0 and (p[i-1] == s[j] or p[i-1] == '.')):
                oneOrMore = self.f(i, j - 1, p, s, dp)

            dp[i][j] = zero or oneOrMore
            return dp[i][j]

        dp[i][j] = False
        return dp[i][j]

    def isMatch(self, s: str, p: str) -> bool:
        n = len(p)
        m = len(s)
        if (n == 0 and m == 0): return True
        if (n == 0): return False

        dp = [[-1]*m for _ in range(n)]
        return self.f(n - 1, m - 1, p, s, dp)

File: 0010_Regular_Expression_Matching/Regular_Expression_Matching.cpp
// TODO: Solve this problem again. Without the help of AI. 

class Solution {
public:
    bool f(int i, int j, string &p, string &s, vector<vector<int>> &dp){
        // base cases
        if (i < 0 && j < 0) return true;
        if (i < 0 && j >= 0) return false;

        // If pattern left but string is empty -> pattern must be x* x* x* ...
        if (i >= 0 && j < 0) {
            while (i >= 0) {
                // '*' must be at i and it must have a preceding char at i-1
                if (p[i] != '*') return false;
                i -= 2;
            }
            return true;
        }

        if (dp[i][j] != -1) return dp[i][j];

        // direct match or '.'
        if (p[i] == s[j] || p[i] == '.') {
            return dp[i][j] = f(i-1, j-1, p, s, dp);
        }

        // star handling: p[i] == '*'
        if (p[i] == '*') {
            // '*' cannot be standalone: need a preceding char at i-1
            // Option 1: zero occurrences of preceding char => skip preceding char and '*'
            bool zero = false;
            if (i - 2 >= -1) // safe to call f(i-2,...); base handles i<0
                zero = f(i - 2, j, p, s, dp);

            // Option 2: one or more occurrences => preceding char must match s[j]
            bool oneOrMore = false;
            if (i - 1 >= 0 && j >= 0 && (p[i-1] == s[j] || p[i-1] == '.')) {
                oneOrMore = f(i, j - 1, p, s, dp);
            }

            return dp[i][j] = zero || oneOrMore;
        }

        return dp[i][j] = false;
    }

    bool isMatch(string s, string p) {
        int n = p.size();
        int m = s.size();
        if (n == 0 && m == 0) return true;
        if (n == 0) return false;

        // If either is zero-length, dp vector dimensioning below must be safe.
        // We'll only index dp[i][j] when i>=0 and j>=0, so vector sizes of n x m are fine.
        vector<vector<int>> dp(n, vector<int>(m, -1));
        return f(n - 1, m - 1, p, s, dp);
    }
};
